crates/subspace-service/src/mmr/sync.rs

```diff
@@ -137,18 +202,45 @@ where
                         // Save the MMR-nodes from response to the local storage
                         'data: for (position, data) in response.mmr_data.iter() {
                             // Ensure continuous sync
-                            if *position == starting_position {
-                                let canon_key = get_offchain_key((*position).into());
-                                offchain_db.local_storage_set(
-                                    StorageKind::PERSISTENT,
-                                    &canon_key,
-                                    data,
+                            if *position != starting_position {
+                                debug!(
+                                    ?peer_info,
+                                    %starting_position,
+                                    %position,
+                                    "MMR sync error: incorrect starting position."
                                 );
 
-                                starting_position += 1;
-                            } else {
-                                debug!("MMR-sync gap detected={peer_id}, position={position}",);
-                                break 'data; // We don't support gaps in MMR data
+                                continue 'peers;
+                            }
+
+                            let node = decode_mmr_data(data);
+
+                            let node = match node {
+                                Ok(node) => node,
+                                Err(err) => {
+                                    debug!(?peer_info, ?err, %position, "Can't decode MMR data received from the peer.");
+
+                                    continue 'peers;
+                                }
+                            };
+
+                            if matches!(node, Node::Data(_)) {
+                                if let Err(err) = mmr.push(node) {
+                                    debug!(?peer_info, ?err, %position, "Can't add MMR data received from the peer.");
+
+                                    return Err(sp_blockchain::Error::Backend(
+                                        "Can't add MMR data to the MMR storage".to_string(),
+                                    ));
+                                }
+
+                                leaves_number += 1;
+                            }
+
+                            starting_position += 1;
+
+                            if u64::from(*position) >= target_position {
```

Comment:
It took me a while to understand the logic here. I don’t think there’s an actual bug, but maybe the loop conditions could be rewritten to be clearer?

- we check `position != starting_position` and only get to this code if they’re equal, then add one to `starting_position`, then compare `position >= target_position`, which guarantees `target_position <= starting_position` below (because that was true when they were equal, before we added one). Is there a simpler way to express this condition that makes this relationship clearer? Maybe a boolean `reached_target` flag or something?
- this condition is `position >= target_position`, but the one below is `target_position <= starting_position`. A consistent order would make it easier to see they’re the same condition.

Totally up to you, but it might also be clearer to split one or more of the inner loops into their own function.